src/Cassandra/Cluster.cs

```diff
@@ -213,8 +213,30 @@ private async Task Init()
 
                     // Only abort the async operations when at least twice the time for ConnectTimeout per host passed
                     var initialAbortTimeout = Configuration.SocketOptions.ConnectTimeoutMillis * 2 * _metadata.Hosts.Count;
-                    initialAbortTimeout = Math.Max(initialAbortTimeout, ControlConnection.MetadataAbortTimeout);
-                    await _controlConnection.InitAsync().WaitToCompleteAsync(initialAbortTimeout).ConfigureAwait(false);
+                    initialAbortTimeout = Math.Max(initialAbortTimeout, Configuration.SocketOptions.MetadataAbortTimeout);
+                    var initTask = _controlConnection.InitAsync();
+                    try
+                    {
+                        await initTask.WaitToCompleteAsync(initialAbortTimeout).ConfigureAwait(false);
+                    }
+                    catch (TimeoutException ex)
+                    {
+                        var newEx = new TimeoutException(
+                            "Cluster initialization was aborted after timing out. This mechanism is put in place to" +
+                            " avoid blocking the calling thread forever. This usually caused by a networking issue" +
+                            " between the client driver instance and the cluster. You can increase this timeout via " +
+                            "the SocketOptions.ConnectTimeoutMillis config setting. This can also be related to deadlocks " +
+                            "caused by mixing synchronous and asynchronous code.", ex);
+                        _initException = new CachedInitErrorException(newEx);
+                        initTask.ContinueWith(t =>
```

Comment:
I'm not sure I understand this continuation.

Reply:
When the init failure is caused by an `OperationTimedOutException`, currently we don't actually handle the result of the task so even if the init task returns an exception with more details about the failure, we just keep throwing the `OperationTimedOutException`. 

This continuation replaces the "cached" exception with the one that the init task returns. Another option would be to just make this continuation log the exception and keep throwing the timed out exception but I decided to go with the exception replacement approach since the single log message is very likely to be hidden in all the log spam that will inevitably happen if an application continuously calls `Connect` on a failed cluster.

On the next major we should change this so that there is no cached exceptions and no unrecoverable state https://datastax-oss.atlassian.net/browse/CSHARP-698

Reply:
ah, makes sense.